lib/Sema/TypeCheckDecl.cpp

```diff
@@ -1496,7 +1482,7 @@ static void checkTypeAccessibilityImpl(
   if (!typeAccessScope.hasValue())
     return;
 
-  if (!*typeAccessScope || *typeAccessScope == contextAccessScope)
+  if (typeAccessScope->isPublic() || *typeAccessScope == contextAccessScope)
```

Comment:
This one's definitely checking equal DCs rather than identical scope. I see what you mean about `==` now.